<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Package for inferring the dynamics of B cell state transitions using B-cell native processes (class-switch recombination, somatic hypermutation) from single-cell RNA sequencing data.">
<title>Single-cell Inference of Class Switch Recombination • sciCSR</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Single-cell Inference of Class Switch Recombination">
<meta property="og:description" content="Package for inferring the dynamics of B cell state transitions using B-cell native processes (class-switch recombination, somatic hypermutation) from single-cell RNA sequencing data.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">sciCSR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Basic sciCSR usage</h6>
    <a class="dropdown-item" href="articles/csr.html">Analysing Class-Switch Recombination in B cell scRNA-seq data using sciCSR</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Comparing CSR/SHM/RNA velocity</h6>
    <a class="dropdown-item" href="articles/comparison.html">Comparing state transitions inferred using CSR, SHM and RNA velocity</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://josephng-bio.org">Joseph Ng's website</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Fraternalilab/sciCSR/" aria-label="github">
    <span class="fa fa-github"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/JosefNg1" aria-label="Twitter">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="scicsr-single-cell-inference-of-class-switch-recombination">sciCSR (Single-cell Inference of Class-Switch Recombination)<a class="anchor" aria-label="anchor" href="#scicsr-single-cell-inference-of-class-switch-recombination"></a>
</h1></div>
<p><img src="reference/figures/sciCSR_logo.png" alt="logo" width="200" height="250"></p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><code>sciCSR</code> is a R package designed to analyse class-switch recombination (CSR) in B cell single-cell RNA sequencing (scRNA-seq) data. CSR is a major process in B cell maturation whereby B cell changes the constant region (“isotype”) of their B cell receptor (BCR) in order to adapt their function to different tissue and biological contexts. Along with somatic hypermutation (SHM) where mutations are accumulated in the variable regions of BCR, CSR is a hallmark of the transitions of B cells from a naive state to acquire memory against antigens. In sciCSR we provides routines to extract information on CSR and SHM from scRNA-seq data and, if available, scBCR-seq (i.e. single-cell BCR repertoire profiling) data.</p>
<p>In theory CSR and SHM information can be used as alternatives to the popular <a href="https://doi.org/10.1371/journal.pcbi.1010492" class="external-link">RNA velocity</a> method to infer transitions between B cells. sciCSR takes forward the extracted CSR/SHM information and inputs them to the <a href="https://cellrank.readthedocs.io/en/stable/" class="external-link">CellRank</a> method to infer transitions between B cells; this can be used directly to analyse patterns of B cell maturation in the data, by considering evidence from CSR and SHM, both of which are native to B cell biology, addressing the <a href="https://doi.org/10.15252/msb.202110282" class="external-link">limitations</a> of RNA velocity in analysing scRNA-seq datasets of mature cell types such as B cells found in circulation or in secondary lymphoid organs.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p><strong>Note:</strong> <em>the installation and use of the sciCSR package has been tested in Linux, MacOS and Windows machines. That said, there may still be uncatched issues in installation and/or usage - please log an issue in the <a href="https://github.com/Fraternalilab/sciCSR/" class="external-link">github</a> and we will look into them.</em></p>
<pre><code><span><span class="co"># install dependencies and sciCSR itself (requires 'devtools')</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">'devtools'</a></span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'devtools'</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mojaveazure/seurat-disk"</span>, <span class="st">"Fraternalilab/sciCSR"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre>
<p>sciCSR requires a few python packages to be installed (<a href="https://scanpy.readthedocs.io/" class="external-link">scanpy</a>, <a href="https://scvelo.readthedocs.io/en/stable/" class="external-link">scvelo</a>, cellrank etc.) as its functionalities depend on these python packages. We recommend setting up a <a href="https://docs.conda.io/projects/conda/en/latest/user-guide/concepts/environments.html" class="external-link">conda environment</a> with these dependencies so that sciCSR exclusively calls packages from this environment; this will avoid problems with messing up your local installation of python (if any) if you have previously set up these packages for other uses. You don’t need to externally set this up - <strong>before the first time you run any examples/analyses with sciCSR, just run the following lines in R</strong> after finishing the sciCSR installation:</p>
<pre class="{r}"><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">sciCSR</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/prepare_sciCSR.html">prepare_sciCSR</a></span><span class="op">(</span><span class="op">)</span></span></code></pre>
<p>This will set up an environment named ‘scicsr’ with all the desired dependencies ready for your analysis. The code should detect any existing Anaconda/Miniconda installation and uses that to set up the environment; if not, it will install Miniconda at the R default location and set up the conda environment.</p>
</div>
<div class="section level2">
<h2 id="examples">Examples<a class="anchor" aria-label="anchor" href="#examples"></a>
</h2>
<p>Please consult the following vignettes:</p>
<ul>
<li>
<a href="articles/csr.html"><strong>Basic sciCSR usage - Analysing CSR</strong></a>: This vignette analyses a subset of data from <a href="https://doi.org/10.4049/jimmunol.2000280" class="external-link">Hong et al. J Immunol 2020</a> analysing splenic B cells with Il23 knockout which biases B cells away from switches towards the IgG2b isotype.</li>
<li>
<a href="articles/comparison.html"><strong>Comparing cell state transitions inferred using RNA velocity, CSR and SHM</strong></a>: This vignette showcases the use of CSR and SHM information to infer B cell state transitions, and compares them to RNA velocity. A subset of tonsillar B cell scRNA-seq data from <a href="https://doi.org/10.1126/sciimmunol.abe6291" class="external-link">King et al. Sci Immunol 2021</a> was used as example.</li>
</ul>
</div>
</div>
  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing sciCSR</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Joseph Ng <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Joseph Ng.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
