<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="sciCSR">
<title>Comparing state transitions inferred using CSR, SHM and RNA velocity â€¢ sciCSR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Comparing state transitions inferred using CSR, SHM and RNA velocity">
<meta property="og:description" content="sciCSR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">sciCSR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Basic sciCSR usage</h6>
    <a class="dropdown-item" href="../articles/csr.html">Analysing Class-Switch Recombination in B cell scRNA-seq data using sciCSR</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Comparing CSR/SHM/RNA velocity</h6>
    <a class="dropdown-item" href="../articles/comparison.html">Comparing state transitions inferred using CSR, SHM and RNA velocity</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://josephng-bio.org">Joseph Ng's website</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Fraternalilab/sciCSR/" aria-label="github">
    <span class="fa fa-github"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/JosefNg1" aria-label="Twitter">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Comparing state transitions inferred using CSR, SHM and RNA velocity</h1>
            
      
      
      <div class="d-none name"><code>comparison.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In this vignette we will show you how to compare the transitions
inferred using CSR and SHM using sciCSR, against those based on RNA
velocity. You are advised to follow the <a href="csr.html">Basic sciCSR
usage vignette</a> first to understand the basic functionalities of
sciCSR. Here, we will use sciCSR to infer state transitions from B cell
scRNA-seq data based on CSR and SHM information, utilise the same
framework to extrapolate transitions based on RNA velocity, and compare
these inferences visually.</p>
<p>Start by loading the sciCSR package. <strong>Before you run any
examples/analysis with sciCSR for the FIRST time</strong>, run the
<code><a href="../reference/prepare_sciCSR.html">prepare_sciCSR()</a></code> function - this will install all python
dependencies for sciCSR to function.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">sciCSR</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co"># We also need the Seurat package as sciCSR interfaces </span></span>
<span><span class="co"># with Seurat data objects to manipulate scRNA-seq data</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run this if this is the first time you set up sciCSR on </span></span>
<span><span class="co"># your machine; this set up the dependencies in python</span></span>
<span><span class="fu"><a href="../reference/prepare_sciCSR.html">prepare_sciCSR</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "All dependencies have been installed. Have fun with running sciCSR!"</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="load-the-dataset">Load the dataset<a class="anchor" aria-label="anchor" href="#load-the-dataset"></a>
</h2>
<p>In this example we will utilise B cell scRNA-seq data from tonsil
samples taken from <a href="https://doi.org/10.1126/sciimmunol.abe6291" class="external-link">King et al.</a>. Part
of this data (from the donor BCP4) is included in the sciCSR
package:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">king</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/King_Tonsils_BCP4.rds"</span>, package <span class="op">=</span> <span class="st">"sciCSR"</span><span class="op">)</span></span>
<span><span class="va">king</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">king</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">king</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 320 features across 781 samples within 2 assays </span></span>
<span><span class="co">## Active assay: RNA (296 features, 274 variable features)</span></span>
<span><span class="co">##  1 other assay present: IGHC</span></span>
<span><span class="co">##  2 dimensional reductions calculated: umap, pca</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot a UMAP projection</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">king</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="comparison_files/figure-html/unnamed-chunk-2-1.png" width="576"></p>
<p>The dataset was downloaded from <a href="https://www.ebi.ac.uk/biostudies/arrayexpress/studies/E-MTAB-9005?accession=E-MTAB-9005" class="external-link">ArrayExpress</a>;
it came in the form of a Seurat data object where SHM information from
scBCR-seq (i.e.Â single-cell BCR repertoire profiling) data was already
added as cell metadata. For your convenience, we have used sciCSR to
generate a count matrix for productive and sterile heavy-chain
immunoglobulin transcripts, and attached it to the Seurat object. Check
out the <a href="csr.html">basic sciCSR vignette</a> if you need any
help on this analysis on your own datasets. Here, since we are going to
use CSR and SHM information to infer transitions, letâ€™s make sure we
have these in place:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># SHM is already included as a column in the metadata ('IGH_MU_FREQ')</span></span>
<span><span class="co"># so is the BCR isotype from repertoire sequencing ('ISOTYPE')</span></span>
<span></span>
<span><span class="co"># calculate CSR potential (if you need help on this make sure you</span></span>
<span><span class="co"># go through the basic sciCSR usage vignette first)</span></span>
<span><span class="va">king</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getCSRpotential.html">getCSRpotential</a></span><span class="op">(</span></span>
<span>  SeuratObj <span class="op">=</span> <span class="va">king</span>, reference_based <span class="op">=</span> <span class="st">"human"</span>, </span>
<span>  c_gene_anno_name <span class="op">=</span> <span class="st">"ISOTYPE"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># We will use 'IGH_MU_FREQ' as the SHM potential</span></span>
<span><span class="co"># Since sciCSR expects a column 'shm' which indicates this,</span></span>
<span><span class="co"># create a column with this name</span></span>
<span><span class="co"># and just replace NA with 0 as it will cause issues with NA</span></span>
<span><span class="va">king</span><span class="op">$</span><span class="va">shm</span>  <span class="op">&lt;-</span> <span class="va">king</span><span class="op">$</span><span class="va">IGH_MU_FREQ</span></span>
<span><span class="va">king</span><span class="op">$</span><span class="va">shm</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">king</span><span class="op">$</span><span class="va">shm</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="rna-velocity-analysis">RNA velocity analysis<a class="anchor" aria-label="anchor" href="#rna-velocity-analysis"></a>
</h2>
<p>RNA velocity analysis is typically performed using the BAM files as
input with the python packages <a href="http://velocyto.org/velocyto.py/index.html" class="external-link">velocyto</a> and <a href="https://scvelo.readthedocs.io/en/stable/" class="external-link">scVelo</a>. We <a href="http://velocyto.org/velocyto.py/tutorial/cli.html" class="external-link">usually</a>
start with velocyto to generate a <code>.loom</code> file which contains
quantification of spliced and unspliced transcripts, and then used
packages such as scVelo to calculate RNA velocity and generate
visualisations (e.g.Â plot streams of arrows on top of UMAP projection to
depict the direction of flow based on RNA velocity.</p>
<p>In sciCSR, we expect users to have run velocyto on the input BAM
files, but we provide functions in R that interface with scVelo at the
backend to perform RNA velocity calculations. As such, you can stay
within R to perform these analyses!</p>
<p>We have made avaialble 2 <code>.loom</code> files corresponds to the
scRNA-seq data shown above. They are too large to be included in the
sciCSR package itself, but you can download them and try this out
yourself:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">loom_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"https://josephng-bio.org/assets/suppdata/BCP004_Total.loom"</span>,</span>
<span>  <span class="st">"https://josephng-bio.org/assets/suppdata/BCP004_MBC.loom"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">file</span> <span class="kw">in</span> <span class="va">loom_files</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="va">file</span>, destfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/basename.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">file</span><span class="op">)</span>, mode <span class="op">=</span> <span class="st">'wb'</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Here, sequencing reads from cells from the BCP004 samples come from
two sequencing libraries, one specifically from the memory B cells (MBC)
and another from total B cells. The <code>.loom</code> files contain
counts of unspliced and spliced transcripts from each cell represented
in the two libraries. Our goal is first to merge these two count
matrices with the standard gene expression counts stored in the Seurat
data object, and then pass this merged data to scVelo to calculate RNA
velocity.</p>
<p>sciCSR offers functions to merge these <code>.loom</code> files into
one and subsequently to the Seurat object, and then export this for
scVeloâ€™s use by means of a python <a href="https://anndata.readthedocs.io/en/latest/" class="external-link">AnnData</a> object. But
first, letâ€™s have a deeper look into what exactly are the outstanding
issues that sciCSR is trying to solve in this task.</p>
<p>We can actually see the sequencing library information reflected in
the Seurat data object:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># There is a column 'Sample' in the Seurat object that contains</span></span>
<span><span class="co"># this information</span></span>
<span><span class="co"># We can show this by colours on the UMAP</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">king</span>, group.by <span class="op">=</span> <span class="st">"Sample"</span><span class="op">)</span></span></code></pre></div>
<p><img src="comparison_files/figure-html/unnamed-chunk-5-1.png" width="576"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># This is also reflected in the cell barcodes</span></span>
<span><span class="co"># Cells from the MBC library</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Cells.html" class="external-link">Cells</a></span><span class="op">(</span><span class="va">king</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">king</span><span class="op">$</span><span class="va">Sample</span> <span class="op">==</span> <span class="st">"BCP4_MBC"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "BCP4_MBC_AAACCTGCATAGTAAG" "BCP4_MBC_AAAGCAAGTACCAGTT"</span></span>
<span><span class="co">## [3] "BCP4_MBC_AAAGCAAGTAGGCTGA" "BCP4_MBC_AACACGTAGATGTAAC"</span></span>
<span><span class="co">## [5] "BCP4_MBC_AACTCAGCAAGTAATG" "BCP4_MBC_AACTGGTAGAGACGAA"</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Cells from the Total library</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Cells.html" class="external-link">Cells</a></span><span class="op">(</span><span class="va">king</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">king</span><span class="op">$</span><span class="va">Sample</span> <span class="op">==</span> <span class="st">"BCP4_Total"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "BCP4_Total_AAATGCCTCGCCTGAG" "BCP4_Total_AACACGTTCTAACGGT"</span></span>
<span><span class="co">## [3] "BCP4_Total_AACCATGAGAATTGTG" "BCP4_Total_AACCGCGGTCAGAGGT"</span></span>
<span><span class="co">## [5] "BCP4_Total_AACCGCGGTTACCGAT" "BCP4_Total_AACTCAGCAAGCCCAC"</span></span></code></pre>
<p>We have prefixes <code>BCP4_MBC_</code> and <code>BCP4_Total_</code>
appended to the cell barcodes. If we inspect the two <code>.loom</code>
files separately, this isnâ€™t the case:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the function 'read.loom.matrices' simplified fron the one in the </span></span>
<span><span class="co"># velocyto.R package can read each loom matrix separately</span></span>
<span><span class="co"># here read one just to see what's in there</span></span>
<span><span class="va">mbc_loom</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_loom_matrices.html">read_loom_matrices</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/basename.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">loom_files</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">mbc_loom</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           Length   Class     Mode</span></span>
<span><span class="co">## spliced   39272998 dgCMatrix S4  </span></span>
<span><span class="co">## unspliced 39272998 dgCMatrix S4  </span></span>
<span><span class="co">## ambiguous 39272998 dgCMatrix S4</span></span></code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the loom file basically contain 3 matrices, one for spliced transcript </span></span>
<span><span class="co"># counts, one for unspliced and one for ambiguous (i.e. it can't tell</span></span>
<span><span class="co"># whether it is spliced or not)</span></span>
<span><span class="co"># just print the top corner of the spliced matrix here</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">mbc_loom</span><span class="op">[[</span><span class="st">"spliced"</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 3 x 3 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">##            BCP004_Total:AAAGATGCACTAGTAC BCP004_Total:AAAGATGGTTGTCGCG</span></span>
<span><span class="co">## FAM138A                                .                             .</span></span>
<span><span class="co">## AL627309.1                             .                             .</span></span>
<span><span class="co">## AL627309.3                             .                             .</span></span>
<span><span class="co">##            BCP004_Total:AAAGCAACACGGATAG</span></span>
<span><span class="co">## FAM138A                                .</span></span>
<span><span class="co">## AL627309.1                             .</span></span>
<span><span class="co">## AL627309.3                             .</span></span></code></pre>
<p>In this case, although for cell barcodes are also appended with a
prefix, the formats of these are different: in the Seurat object, it is
<code>BCP4_Total_</code>, here it is <code>BCP004_Total:</code> (notice
â€˜BCP004â€™ versus â€˜BCP4â€™, and the underscore/colon at the trailing end).
This will cause problems with merging if not addressed. Luckily, in
sciCSR we offer a function <code><a href="../reference/combineLoomFiles.html">combineLoomFiles()</a></code> which allows
you to combine separate loom files into one, while checking for these
discrepancies of cell barcodes and try to correct them. You will need to
pass the Seurat object as one of the arguments to call this function,
while at the same time give a list of sample names, in the <em>same
order</em> as the vector of loom files you supply - see below:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># merging loom files</span></span>
<span><span class="fu"><a href="../reference/combineLoomFiles.html">combineLoomFiles</a></span><span class="op">(</span></span>
<span>  loom_files <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"BCP004_Total.loom"</span>, <span class="st">"BCP004_MBC.loom"</span><span class="op">)</span>,</span>
<span>  new_loom_filename <span class="op">=</span> <span class="st">'King_Tonsils_BCP4_combined.loom'</span>,</span>
<span>  SeuratObj <span class="op">=</span> <span class="va">king</span>, </span>
<span>  <span class="co"># the order below corresponds to the order in the argument 'loom_files'</span></span>
<span>  sample_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"BCP4_Total"</span>, <span class="st">"BCP4_MBC"</span><span class="op">)</span>, </span>
<span>  seurat_sample_column <span class="op">=</span> <span class="st">"Sample"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Assuming the order in sample_names correspond to the order in loom_files. If this is not the case please rerun this function ensuring the order of these vectors match up.</span></span></code></pre>
<pre><code><span><span class="co">## [1] "Loom object with velocyto spliced/unspliced counts written to file 'King_Tonsils_BCP4_combined.loom'."</span></span></code></pre>
<p>We are next ready to merge these unspliced and spliced counts to the
annotated data object. Since we are going to rely on the python pacakge
scVelo for RNA velocity analysis, here we convert the Seurat object into
an AnnData file, and then merge the velocyto data into this AnnData
object - we provide two separate functions,
<code><a href="../reference/convertSeuratToH5ad.html">convertSeuratToH5ad()</a></code> and
<code><a href="../reference/mergeVelocytoWithGEX.html">mergeVelocytoWithGEX()</a></code>.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># convert the Seurat object into AnnData</span></span>
<span><span class="fu"><a href="../reference/convertSeuratToH5ad.html">convertSeuratToH5ad</a></span><span class="op">(</span></span>
<span>  SeuratObj <span class="op">=</span> <span class="va">king</span>, assays <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>  h5ad_filename <span class="op">=</span> <span class="st">"King_Tonsils_BCP4.h5ad"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "King_Tonsils_BCP4_assay-RNA.h5ad"</span></span></code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># add velocyto loom data into AnnData</span></span>
<span><span class="fu"><a href="../reference/mergeVelocytoWithGEX.html">mergeVelocytoWithGEX</a></span><span class="op">(</span></span>
<span>  anndata_file <span class="op">=</span> <span class="st">"King_Tonsils_BCP4_assay-RNA.h5ad"</span>,</span>
<span>  loom_file <span class="op">=</span> <span class="st">'King_Tonsils_BCP4_combined.loom'</span>,</span>
<span>  anndata_out_filename <span class="op">=</span> <span class="st">'King_Tonsils_BCP4_assay-RNA_velocyto.h5ad'</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "AnnData file with velocyto counts merged is written to King_Tonsils_BCP4_assay-RNA_velocyto.h5ad"</span></span></code></pre>
<p>This file <code>King_Tonsils_BCP4_assay-RNA_velocyto.h5ad</code>
contains the unspliced/spliced transcript counts which can be used as
input to the scVelo RNA velocity calculation pipeline. In sciCSR we
provide a R function <code><a href="../reference/run_scVelo.html">run_scVelo()</a></code> that calls scVelo at the
backend to perform this calculation - all you need to do is to supply
this file and a new filename where the calculated RNA velocity
statistics will be stored.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># split into wt and kd</span></span>
<span><span class="fu"><a href="../reference/run_scVelo.html">run_scVelo</a></span><span class="op">(</span></span>
<span>  anndata_file <span class="op">=</span> <span class="st">"King_Tonsils_BCP4_assay-RNA_velocyto.h5ad"</span>,</span>
<span>  anndata_out_filename <span class="op">=</span> <span class="st">"King_Tonsils_BCP4_assay-RNA_velocyto_scVelo.h5ad"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: The request to</span></span>
<span><span class="co">## `use_python("/home/josephn/anaconda3/envs/scicsr/bin/python")` will be ignored</span></span>
<span><span class="co">## because the environment variable RETICULATE_PYTHON is set to</span></span>
<span><span class="co">## "/home/josephn/anaconda3/envs/scicsr/bin/python3.9"</span></span></code></pre>
<p>We can project the RNA velocity results onto the UMAP using the
plotting functionalities within scVelo. Once again, in sciCSR you can do
this by calling a function in R without the need to code in python -
sciCSR calls scVelo at the backend and return you the plot within R:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_arrows.html">plot_arrows</a></span><span class="op">(</span></span>
<span>  anndata_file <span class="op">=</span> <span class="st">"King_Tonsils_BCP4_assay-RNA_velocyto_scVelo.h5ad"</span>, </span>
<span>  <span class="co"># we are using RNA velocity information to project arrows</span></span>
<span>  based_on <span class="op">=</span> <span class="st">"velocity"</span>,</span>
<span>  <span class="co"># we want to indicate cell cluster (in the column 'CellType' </span></span>
<span>  <span class="co"># of the metadata) using different colours</span></span>
<span>  colour.by <span class="op">=</span> <span class="st">"CellType"</span>,</span>
<span>  <span class="co"># you can indicate a comma-separated list of colours </span></span>
<span>  <span class="co"># each corresponding to a CellType (this is optional)</span></span>
<span>  cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#F8766D"</span>, <span class="st">"#DE8C00"</span>, <span class="st">"#B79F00"</span>, <span class="st">"#7CAE00"</span>, <span class="st">"#00BA38"</span>, </span>
<span>           <span class="st">"#00C08B"</span>, <span class="st">"#00BFC4"</span>, <span class="st">"#00B4F0"</span>, <span class="st">"#619CFF"</span>, <span class="st">"#C77CFF"</span>,</span>
<span>           <span class="st">"#F564E3"</span>, <span class="st">"#FF64B0"</span><span class="op">)</span>,</span>
<span>  style <span class="op">=</span> <span class="st">"stream"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: The request to</span></span>
<span><span class="co">## `use_python("/home/josephn/anaconda3/envs/scicsr/bin/python")` will be ignored</span></span>
<span><span class="co">## because the environment variable RETICULATE_PYTHON is set to</span></span>
<span><span class="co">## "/home/josephn/anaconda3/envs/scicsr/bin/python3.9"</span></span></code></pre>
<p><img src="comparison_files/figure-html/unnamed-chunk-10-1.png" width="576"></p>
<p>We now have a RNA velocity plot based on code entirely in R (sciCSR
does all the magic in python in the backend)! We can also plot RNA
velocity as a grid of arrows rather than as arrow streams as above:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_arrows.html">plot_arrows</a></span><span class="op">(</span></span>
<span>  anndata_file <span class="op">=</span> <span class="st">"King_Tonsils_BCP4_assay-RNA_velocyto_scVelo.h5ad"</span>, </span>
<span>  <span class="co"># we are using RNA velocity information to project arrows</span></span>
<span>  based_on <span class="op">=</span> <span class="st">"velocity"</span>,</span>
<span>  <span class="co"># we want to indicate cell cluster (in the column 'CellType' </span></span>
<span>  <span class="co"># of the metadata) using different colours</span></span>
<span>  colour.by <span class="op">=</span> <span class="st">"CellType"</span>,</span>
<span>  <span class="co"># you can indicate a comma-separated list of colours </span></span>
<span>  <span class="co"># each corresponding to a CellType (this is optional)</span></span>
<span>  cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#F8766D"</span>, <span class="st">"#DE8C00"</span>, <span class="st">"#B79F00"</span>, <span class="st">"#7CAE00"</span>, <span class="st">"#00BA38"</span>, </span>
<span>           <span class="st">"#00C08B"</span>, <span class="st">"#00BFC4"</span>, <span class="st">"#00B4F0"</span>, <span class="st">"#619CFF"</span>, <span class="st">"#C77CFF"</span>,</span>
<span>           <span class="st">"#F564E3"</span>, <span class="st">"#FF64B0"</span><span class="op">)</span>,</span>
<span>  <span class="co"># indicate 'grid' if you want to plot arrows onto a grid of fixed points</span></span>
<span>  style <span class="op">=</span> <span class="st">"grid"</span>,</span>
<span>  <span class="co"># (Optional) you can give the plot a customised title as well</span></span>
<span>  title <span class="op">=</span> <span class="st">"King et al. Donor BCP4 RNA velocity on a grid"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: The request to</span></span>
<span><span class="co">## `use_python("/home/josephn/anaconda3/envs/scicsr/bin/python")` will be ignored</span></span>
<span><span class="co">## because the environment variable RETICULATE_PYTHON is set to</span></span>
<span><span class="co">## "/home/josephn/anaconda3/envs/scicsr/bin/python3.9"</span></span></code></pre>
<p><img src="comparison_files/figure-html/unnamed-chunk-11-1.png" width="576"></p>
</div>
<div class="section level2">
<h2 id="plotting-transitions-based-on-csr-and-shm">Plotting transitions based on CSR and SHM<a class="anchor" aria-label="anchor" href="#plotting-transitions-based-on-csr-and-shm"></a>
</h2>
<p>Since sciCSR infers transitions using CSR and/or SHM information, we
would ideally want to also visualise these, and compare these inferred
transitions against those based on RNA velocity. We can do so by
projecting the transitions onto UMAP similar to what is shown above for
RNA velocity, thanks to functionalities support in the CellRank package.
Before we show you these, however, some points of caution:</p>
<p>Inferred transitions based on CSR and/or SHM can be substantially
different from those based on RNA velocity. Specifically, it is fairly
conceivable that depending on the dataset, even using CSR/SHM <em>we
will not see much prominent transitional flows</em>. Part of it may be
genuine (e.g.Â you are sampling mainly mature B cells at a steady state
with little difference in CSR/SHM patterns) - in fact, it has been
noticed that one <a href="https://doi.org/10.15252/msb.202110282" class="external-link">common
problem with RNA velocity</a> is the prominence of flows (sometimes even
pointing to a counterintuitive direction) in mature immune cell types
where little dynamical behaviour is expected. Another potential reason
for the lack of prominent flows in these sort of visualisation is,
<em>cells are not necessarily organised by the states defined using
CSR/SHM information on the UMAP</em>. For example, IgG1+ and IgA1+ B
cells may occupy similar positions on a UMAP projection simply because
they share transcriptomic similarities outside of the immunoglobulin
loci. In this case, even it is conceivable that there can be a G1-to-A1
transition, projection of arrows onto the UMAP reduction isnâ€™t
particularly helpful as we may not have these states positioning on the
UMAP in a way that we can highlight transitions between them using the
arrows.</p>
<p>Anyway, enough health warning - we can also use the
<code><a href="../reference/plot_arrows.html">plot_arrows()</a></code> function to project CSR/SHM information onto
the UMAP. Just indicate this in the <code>based_on</code> argument:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Let's say we want to first project CSR-inferred transitions </span></span>
<span><span class="co"># onto the UMAP</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_arrows.html">plot_arrows</a></span><span class="op">(</span></span>
<span>  anndata_file <span class="op">=</span> <span class="st">"King_Tonsils_BCP4_assay-RNA_velocyto_scVelo.h5ad"</span>, </span>
<span>  <span class="co"># allowed values: 'velocity' (default), 'csr' or 'shm'</span></span>
<span>  based_on <span class="op">=</span> <span class="st">"csr"</span>,</span>
<span>  colour.by <span class="op">=</span> <span class="st">"CellType"</span>,</span>
<span>  cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#F8766D"</span>, <span class="st">"#DE8C00"</span>, <span class="st">"#B79F00"</span>, <span class="st">"#7CAE00"</span>, <span class="st">"#00BA38"</span>, </span>
<span>           <span class="st">"#00C08B"</span>, <span class="st">"#00BFC4"</span>, <span class="st">"#00B4F0"</span>, <span class="st">"#619CFF"</span>, <span class="st">"#C77CFF"</span>,</span>
<span>           <span class="st">"#F564E3"</span>, <span class="st">"#FF64B0"</span><span class="op">)</span>,</span>
<span>  style <span class="op">=</span> <span class="st">"stream"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: The request to</span></span>
<span><span class="co">## `use_python("/home/josephn/anaconda3/envs/scicsr/bin/python")` will be ignored</span></span>
<span><span class="co">## because the environment variable RETICULATE_PYTHON is set to</span></span>
<span><span class="co">## "/home/josephn/anaconda3/envs/scicsr/bin/python3.9"</span></span></code></pre>
<p><img src="comparison_files/figure-html/unnamed-chunk-12-1.png" width="576"></p>
<p>And another plot using SHM information:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_arrows.html">plot_arrows</a></span><span class="op">(</span></span>
<span>  anndata_file <span class="op">=</span> <span class="st">"King_Tonsils_BCP4_assay-RNA_velocyto_scVelo.h5ad"</span>, </span>
<span>  based_on <span class="op">=</span> <span class="st">"shm"</span>,</span>
<span>  colour.by <span class="op">=</span> <span class="st">"CellType"</span>,</span>
<span>  cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#F8766D"</span>, <span class="st">"#DE8C00"</span>, <span class="st">"#B79F00"</span>, <span class="st">"#7CAE00"</span>, <span class="st">"#00BA38"</span>, </span>
<span>           <span class="st">"#00C08B"</span>, <span class="st">"#00BFC4"</span>, <span class="st">"#00B4F0"</span>, <span class="st">"#619CFF"</span>, <span class="st">"#C77CFF"</span>,</span>
<span>           <span class="st">"#F564E3"</span>, <span class="st">"#FF64B0"</span><span class="op">)</span>,</span>
<span>  style <span class="op">=</span> <span class="st">"stream"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: The request to</span></span>
<span><span class="co">## `use_python("/home/josephn/anaconda3/envs/scicsr/bin/python")` will be ignored</span></span>
<span><span class="co">## because the environment variable RETICULATE_PYTHON is set to</span></span>
<span><span class="co">## "/home/josephn/anaconda3/envs/scicsr/bin/python3.9"</span></span></code></pre>
<p><img src="comparison_files/figure-html/unnamed-chunk-13-1.png" width="576"></p>
</div>
<div class="section level2">
<h2 id="more-detailed-comparisons">More detailed comparisons<a class="anchor" aria-label="anchor" href="#more-detailed-comparisons"></a>
</h2>
<p>The visualisations above clearly show some similarities (for example,
the transitions from MBC pointing towards the direction of germinal
center [GC]) and differences (e.g.Â strong transitions in the GC based on
RNA velocity appears not backed up with CSR/SHM) between inferences
based on RNA velocity versus CSR/SHM. sciCSR provides further
functionalities to compare these different sets of inferred transitions.
One way to gain further insights would be to apply Transition Path
Theory (TPT) to estimate the amount of fluxes between states - we have
covered TPT in the <a href="csr.html">basic sciCSR usage vignette</a>.
You may want to take a look at the vignette and see how we can do
that.</p>
<p>Here we show you another way to answer the question of â€˜How different
are the transitions inferred using CSR/SHM to those based on RNA
velocityâ€™. In sciCSR we uses CSR/SHM information to fit models of state
transitions using Markov State Models (MSM). One powerful properties of
MSM is, they specify the transition likelihoods between every state in
the data, such that we can â€˜generateâ€™/â€˜simulateâ€™ new transitions as if
we have data from a new cell and follow how it moves across the
landscape of states. Similar MSMs would produce similar pathways of
transitions across this landscape; we can take a note at the frequencies
each state is visited in these simulated pathways, and then derive a
numerical metric to quantify the similarity in terms of the frequencies
of visiting each states in these MSM-simulated paths (Illustration
below).</p>
<p><img src="images/sample_msm.png" alt="Illustration of sampling MSM and comparing inferences" width="260"></p>
<p>In sciCSR, the way we can perform this analysis is, first fit
transition models and obtain separate R objects which store these
models. We can then use the function
<code><a href="../reference/compareTransitionMatrices.html">compareTransitionMatrices()</a></code> to quantify the similarities of
these transition models:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Let's first fit transition models using velocity/CSR/SHM</span></span>
<span><span class="co"># separately using fitTransitionModel()</span></span>
<span></span>
<span><span class="co"># Velocity</span></span>
<span><span class="va">g_velo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitTransitionModel.html">fitTransitionModel</a></span><span class="op">(</span></span>
<span>  anndata_file <span class="op">=</span> <span class="st">"King_Tonsils_BCP4_assay-RNA_velocyto_scVelo.h5ad"</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"velocity"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: The request to</span></span>
<span><span class="co">## `use_python("/home/josephn/anaconda3/envs/scicsr/bin/python")` will be ignored</span></span>
<span><span class="co">## because the environment variable RETICULATE_PYTHON is set to</span></span>
<span><span class="co">## "/home/josephn/anaconda3/envs/scicsr/bin/python3.9"</span></span></code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># CSR</span></span>
<span><span class="va">g_csr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitTransitionModel.html">fitTransitionModel</a></span><span class="op">(</span></span>
<span>  anndata_file <span class="op">=</span> <span class="st">"King_Tonsils_BCP4_assay-RNA_velocyto_scVelo.h5ad"</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"pseudotime"</span>, pseudotime_key <span class="op">=</span> <span class="st">"csr_pot"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: The request to</span></span>
<span><span class="co">## `use_python("/home/josephn/anaconda3/envs/scicsr/bin/python")` will be ignored</span></span>
<span><span class="co">## because the environment variable RETICULATE_PYTHON is set to</span></span>
<span><span class="co">## "/home/josephn/anaconda3/envs/scicsr/bin/python3.9"</span></span></code></pre>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># SHM</span></span>
<span><span class="va">g_shm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitTransitionModel.html">fitTransitionModel</a></span><span class="op">(</span></span>
<span>  anndata_file <span class="op">=</span> <span class="st">"King_Tonsils_BCP4_assay-RNA_velocyto_scVelo.h5ad"</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"pseudotime"</span>, pseudotime_key <span class="op">=</span> <span class="st">"shm"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: The request to</span></span>
<span><span class="co">## `use_python("/home/josephn/anaconda3/envs/scicsr/bin/python")` will be ignored</span></span>
<span><span class="co">## because the environment variable RETICULATE_PYTHON is set to</span></span>
<span><span class="co">## "/home/josephn/anaconda3/envs/scicsr/bin/python3.9"</span></span></code></pre>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">comparison</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compareTransitionMatrices.html">compareTransitionMatrices</a></span><span class="op">(</span></span>
<span>  <span class="co"># these are the transition matrices based on</span></span>
<span>  <span class="co"># velocity/CSR/SHM</span></span>
<span>  matrix_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">g_velo</span><span class="op">$</span><span class="va">transition_matrix</span>,</span>
<span>                     <span class="va">g_csr</span><span class="op">$</span><span class="va">transition_matrix</span>,</span>
<span>                     <span class="va">g_shm</span><span class="op">$</span><span class="va">transition_matrix</span><span class="op">)</span>,</span>
<span>  SeuratObj <span class="op">=</span> <span class="va">king</span>, cells <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/Cells.html" class="external-link">Cells</a></span><span class="op">(</span><span class="va">king</span><span class="op">)</span>,</span>
<span>  <span class="co"># how do we want to group the cells? we will describe</span></span>
<span>  <span class="co"># each step in the simulated pathways using this</span></span>
<span>  <span class="co"># in this case we want to describe pathways in terms</span></span>
<span>  <span class="co"># of what cell type</span></span>
<span>  group.by <span class="op">=</span> <span class="st">'CellType'</span>,</span>
<span>  <span class="co"># We will compare these simulated paths using Kullback-</span></span>
<span>  <span class="co"># Leibler divergence on the frequencies each state is visited</span></span>
<span>  distance_metric <span class="op">=</span> <span class="st">'KL'</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Sampling realisations from Markov chain 1 ...</span></span></code></pre>
<pre><code><span><span class="co">## Sampling realisations from Markov chain 2 ...</span></span></code></pre>
<pre><code><span><span class="co">## Sampling realisations from Markov chain 3 ...</span></span></code></pre>
<pre><code><span><span class="co">## Comparing the list of transition models ...</span></span></code></pre>
<pre><code><span><span class="co">## Metric: 'kullback-leibler' using unit: 'log2'; comparing: 3 vectors.</span></span></code></pre>
<p>The output of this function is a list with two elements: the
<code>sampled_transitions</code> are the frequencies each transition
(i.e.Â in this case, moves from one cell type to another) is visited in
the pathways sampled by considering the transition model fitted using
velocity/CSR/SHM information. Another element of this list (name
<code>distance</code>) is a matrix which shows the difference between
each pair of transition matrix, ordered based on the one given in
<code>matrix_list</code>. The values in this matrix is a divergence
metric (in this case <a href="https://en.wikipedia.org/wiki/Kullback%E2%80%93Leibler_divergence" class="external-link">Kullback-Leibler
divergence</a>) to quantify how different the transitions implicated in
one distance matrix is different from another. The larger this value,
the more different the implied transitions are.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">distance_matrix</span> <span class="op">&lt;-</span> <span class="va">comparison</span><span class="op">$</span><span class="va">distance</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">distance_matrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Velocity"</span>, <span class="st">"CSR"</span>, <span class="st">"SHM"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">distance_matrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Velocity"</span>, <span class="st">"CSR"</span>, <span class="st">"SHM"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># We can plot this matrix as a heatmap</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="va">distance_matrix</span>, display_numbers <span class="op">=</span> <span class="cn">TRUE</span>, number_format <span class="op">=</span> <span class="st">"%.4f"</span><span class="op">)</span></span></code></pre></div>
<p><img src="comparison_files/figure-html/unnamed-chunk-16-1.png" width="384"></p>
<p>Based on this plot, we can see that velocity-based transitions are
substantially more different than those based on CSR/SHM information.
This serves more as an overall metric to quantify the differences
between those transition models, whilst the arrow projection and/or
TPT-based fluxes, both outputs of different sciCSR functions, will allow
you to study further what exactly are the differences in the inferences
made using each type of biological information (velocity/CSR/SHM)!</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Joseph Ng.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
