<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference â€¢ sciCSR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">sciCSR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Basic sciCSR usage</h6>
    <a class="dropdown-item" href="../articles/csr.html">Analysing Class-Switch Recombination in B cell scRNA-seq data using sciCSR</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Comparing CSR/SHM/RNA velocity</h6>
    <a class="dropdown-item" href="../articles/comparison.html">Comparing state transitions inferred using CSR, SHM and RNA velocity</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://josephng-bio.org">Joseph Ng's website</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Fraternalilab/sciCSR/" aria-label="github">
    <span class="fa fa-github"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/JosefNg1" aria-label="Twitter">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="setting-up-dependencies">Setting up dependencies<a class="anchor" aria-label="anchor" href="#setting-up-dependencies"></a></h2>
      
      <p class="section-desc"></p><p>Wrapper to set up python dependencies.</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="prepare_sciCSR.html">prepare_sciCSR()</a></code> 
        </dt>
        <dd>Prepare conda environment</dd>
      </dl></div><div class="section level2">
      <h2 id="scrna-seq-data-processing-and-wrangling">scRNA-seq data processing and wrangling<a class="anchor" aria-label="anchor" href="#scrna-seq-data-processing-and-wrangling"></a></h2>
      
      <p class="section-desc"></p><p>Functions for basic scRNA-seq data processing wrangling, including B-cell-specific functionalities.</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="collapseIntoMetagenes.html">collapseIntoMetagenes()</a></code> 
        </dt>
        <dd>Group gene counts into metagenes to minimise individual differences</dd>
      </dl><dl><dt>
          
          <code><a href="normalise_dimreduce.html">normalise_dimreduce()</a></code> 
        </dt>
        <dd>Normalisation and dimensionality reduction of scRNAseq gene counts</dd>
      </dl><dl><dt>
          
          <code><a href="guessBarcodes.html">guessBarcodes()</a></code> 
        </dt>
        <dd>parse the substring inside a given cell identifier which corresponds to the nucleotide barcode</dd>
      </dl><dl><dt>
          
          <code><a href="repairBarcode.html">repairBarcode()</a></code> 
        </dt>
        <dd>Repairing cell barcodes in a list of data frames/matrices to match the Seurat object</dd>
      </dl><dl><dt>
          
          <code><a href="convertSeuratToH5ad.html">convertSeuratToH5ad()</a></code> 
        </dt>
        <dd>wrapper function to convert Seurat Object to a AnnData .h5ad file</dd>
      </dl><dl><dt>
          
          <code><a href="splitAnnData.html">splitAnnData()</a></code> 
        </dt>
        <dd>Split AnnData object by levels in a specified meta data trait</dd>
      </dl><dl><dt>
          
          <code><a href="read_loom_matrices.html">read_loom_matrices()</a></code> 
        </dt>
        <dd>read loom matrices</dd>
      </dl><dl><dt>
          
          <code><a href="combineLoomFiles.html">combineLoomFiles()</a></code> 
        </dt>
        <dd>Combine velocyto loom files on multiple BAM files into one loom file</dd>
      </dl></div><div class="section level2">
      <h2 id="merging-repertoire-and-scrna-seq-read-counts">Merging Repertoire and scRNA-seq read counts<a class="anchor" aria-label="anchor" href="#merging-repertoire-and-scrna-seq-read-counts"></a></h2>
      
      <p class="section-desc"></p><p>Functions for additing repertoire annotation features as metadata to gene expression count matrices.</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="annotatePairing.html">annotatePairing()</a></code> 
        </dt>
        <dd>Annotate heavy-light chain pairing</dd>
      </dl><dl><dt>
          
          <code><a href="collapseBCR.html">collapseBCR()</a></code> 
        </dt>
        <dd>Collapse the VDJ repertoire data frames by cell barcode</dd>
      </dl><dl><dt>
          
          <code><a href="AddCellMetaToVDJ.html">AddCellMetaToVDJ()</a></code> 
        </dt>
        <dd>Add cell metadata from Seurat object into VDJ data frame</dd>
      </dl><dl><dt>
          
          <code><a href="combineBCR.html">combineBCR()</a></code> 
        </dt>
        <dd>Add annotations from VDJ data frame into the Seurat object</dd>
      </dl></div><div class="section level2">
      <h2 id="csr-shm-based-pseudotime">CSR/SHM-based pseudotime<a class="anchor" aria-label="anchor" href="#csr-shm-based-pseudotime"></a></h2>
      
      <p class="section-desc"></p><p>Functions for extracting and calculating cell pseudotime ordering using CSR and SHM information.</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="scanBam.html">scanBam()</a></code> 
        </dt>
        <dd>Scan reads mapped to a given genomic range</dd>
      </dl><dl><dt>
          
          <code><a href="getIGHmapping.html">getIGHmapping()</a></code> 
        </dt>
        <dd>wrapper function to scan sterile/productive IGH molecules from BAM file</dd>
      </dl><dl><dt>
          
          <code><a href="getIGHreadType.html">getIGHreadType()</a></code> 
        </dt>
        <dd>Deduce type of IGH reads</dd>
      </dl><dl><dt>
          
          <code><a href="getJunctionReads.html">getJunctionReads()</a></code> 
        </dt>
        <dd>extract reads covering splice junctions</dd>
      </dl><dl><dt>
          
          <code><a href="summariseIGHreads.html">summariseIGHreads()</a></code> 
        </dt>
        <dd>cast data frame IGH counts into a matrix</dd>
      </dl><dl><dt>
          
          <code><a href="mergeIgHCountsToSeurat.html">mergeIgHCountsToSeurat()</a></code> 
        </dt>
        <dd>merge IgH productive/sterile transcript count into Seurat Object</dd>
      </dl><dl><dt>
          
          <code><a href="getCSRpotential.html">getCSRpotential()</a></code> 
        </dt>
        <dd>score cells by their Class Switch Recombination (CSR) status</dd>
      </dl><dl><dt>
          
          <code><a href="getSHM.html">getSHM()</a></code> 
        </dt>
        <dd>get somatic hypermutation level</dd>
      </dl><dl><dt>
          
          <code><a href="getIsotype.html">getIsotype()</a></code> 
        </dt>
        <dd>label isotypes based on productive/sterile transcript levels</dd>
      </dl></div><div class="section level2">
      <h2 id="rna-velocity">RNA velocity<a class="anchor" aria-label="anchor" href="#rna-velocity"></a></h2>
      
      <p class="section-desc"></p><p>RNA velocity analysis in R by interfacing with python packages</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="mergeVelocytoWithGEX.html">mergeVelocytoWithGEX()</a></code> 
        </dt>
        <dd>Combine velocyto loom data with AnnData</dd>
      </dl><dl><dt>
          
          <code><a href="run_scVelo.html">run_scVelo()</a></code> 
        </dt>
        <dd>Calculate RNA velocity using the python scvelo workflow</dd>
      </dl></div><div class="section level2">
      <h2 id="inferring-transitions">Inferring transitions<a class="anchor" aria-label="anchor" href="#inferring-transitions"></a></h2>
      
      <p class="section-desc"></p><p>Inferring transitions using CSR/SHM information or RNA velocity</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="fitTransitionModel.html">fitTransitionModel()</a></code> 
        </dt>
        <dd>Fit transition model on data using the python cellrank package</dd>
      </dl><dl><dt>
          
          <code><a href="fitTPT.html">fitTPT()</a></code> 
        </dt>
        <dd>Fit Transition Path Theory (TPT) on the cellrank transition models</dd>
      </dl><dl><dt>
          
          <code><a href="compareTransitionMatrices.html">compareTransitionMatrices()</a></code> 
        </dt>
        <dd>Computing distances between multiple transition matrices</dd>
      </dl></div><div class="section level2">
      <h2 id="data-visualisation">Data visualisation<a class="anchor" aria-label="anchor" href="#data-visualisation"></a></h2>
      
      <p class="section-desc"></p><p>Visualising inferred transitions</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="plotFluxMatrix.html">plotFluxMatrix()</a></code> 
        </dt>
        <dd>Visualise flux matrix describing class-switch recombination (CSR) transitions in data</dd>
      </dl><dl><dt>
          
          <code><a href="plotStationaryDistribution.html">plotStationaryDistribution()</a></code> 
        </dt>
        <dd>Visualise stationary distribution of isotypes in the data</dd>
      </dl><dl><dt>
          
          <code><a href="plot_arrows.html">plot_arrows()</a></code> 
        </dt>
        <dd>Plot arrows showing transitions in the style of RNA velocity plots</dd>
      </dl></div><div class="section level2">
      <h2 id="data-objects">Data objects<a class="anchor" aria-label="anchor" href="#data-objects"></a></h2>
      
      <p class="section-desc"></p><p>Data objects underlying sciCSR functionalities</p>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="human_definitions.html">human_definitions</a></code> 
        </dt>
        <dd>Genomic coordinates of human heavy-chain immunoglobulin V, D, J, C genes</dd>
      </dl><dl><dt>
          
          <code><a href="human_nmf.html">human_nmf</a></code> 
        </dt>
        <dd>Isotype sterile/productive signatures trained using human single-cell B cell atlas</dd>
      </dl><dl><dt>
          
          <code><a href="mouse_definitions.html">mouse_definitions</a></code> 
        </dt>
        <dd>Genomic coordinates of mouse heavy-chain immunoglobulin V, D, J, C genes</dd>
      </dl><dl><dt>
          
          <code><a href="mouse_nmf.html">mouse_nmf</a></code> 
        </dt>
        <dd>Isotype sterile/productive signatures trained using mouse single-cell B cell atlas</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Joseph Ng.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

